### DISCUSSION TITLE
Enabling built in keyboard and trackpad

### DISCUSSION THREAD

**Discussion by smedir** [jglathe/linux_ms_dev_kit#18] [Q&A] [answered]
Thanks for this repository and image. I am installing on a Lenovo Yoga Slim 7X.

I was able to install debian with the installer from Linaro, had working wi fi, and built in keyboard, no trackpad, sound.

I compiled vanilla 6.11-rc4 natively, which works, however no built in keyboard, trackpad, sound. Only an external usb plugged keyboard and mouse work. Using the kernel compiled Lenovo Yoga Slim7X dtb file.

When I boot from your kernel and image, using your initrd, from the default grub menu for the lenovo Yoga which used the dtb file in your image, and only giving the kernel the debian partition as root to the kernel on the command line,  debian comes up, and the built in keyboard and trackpad work perfectly! Even swiping, gestures work.

What exactly in the initrd or the dtb file used to boot that grub entry makes the keyboard and trackpad work? Is it the defconfig used to compile the kernel, firmware in the initrd or the dtb definition? Am I missing loading some specific modules for the the devices?

I would like to use the same method to enable the my compiled vanilla kernel do exactly the same thing.

Thanks for your help

**Comment by smedir** [comment]
Ok, I cannot reproduce it now. Debian Trixie updated, and your kernel crashes the machine after "start plymouth boot screen" service.

For a while your kernel was booting with the keyboard, trackpad, gestures all working. In fact even the touch screen was working!
I don't know how that happened.

I can still boot the vanilla compiled kernel, but no native keyboard or touchpad!

**Reply by smedir** [reply]
This is the kernel dmesg from your kernel when the trackpad, touchscreen and keyboard were working-

[dmesg-success.txt](https://github.com/user-attachments/files/16790980/dmesg-success.txt)


This is the kernel dmesg from the vanilla kernel from kernel.org, with no working native keyboard or trackpad-

[dmesg-nokeyboard.txt](https://github.com/user-attachments/files/16790996/dmesg-nokeyboard.txt)

**Reply by jglathe** [reply]
hmm is evdev in the modules list?

**Comment by jglathe** [comment]
Sorry for the late reply, need to check my comms settings (I got no email that there is a new discussion). Most recent Image is V5, it has evdev listed in the modules to load. Please try this one. Also, V6 is in preparation, with some local testing (finally!) although on a HP Omnibook X14. First boot didn't show internal keyboard either, I'm checking.

**Reply by jglathe** [reply]
the file which loads evdev (and others) is `/etc/initramfs-tools/modules`. Needs to be owned by root and executable (afair) to be used when doing `update-initramfs`

**Reply by smedir** [reply, marked as answer]
It is not evdev.

> On X1E80100, GPIO interrupts for wakeup-capable pins have been broken since
> the introduction of the pinctrl driver. This prevents keyboard and touchpad
> from working on most of the X1E laptops. So far we have worked around this
> by manually building a kernel with the "wakeup-parent" removed from the
> pinctrl node in the device tree, but we cannot expect all users to do that.

[Link to the patch on kernel mailing list](https://lore.kernel.org/all/CACRpkdY97ZOOu66FkLcCSF9vd5uw9N6i239t8aeuaAR4B_ccxA@mail.gmail.com/)

Applying that patch to vanilla rc4 kernel fixed the problem. Trackpad, keyboard and touchscreen.

Your dtb file and the wakeup-parent fix, that is why everything worked.

Your latest rc7 kernel works as well, normal keyboard trackpad and even touchscreen.
